lib/babe/verify.go

```diff
@@ -327,62 +327,58 @@ func (b *verifier) verifyAuthorshipRight(header *types.Header) error {
 		return ErrBadSignature
 	}
 
-	// check if the producer has equivocated, ie. have they produced a conflicting block?
-	// hashes is hashes of all blocks with same block number as header.Number
-	hashes := b.blockState.GetAllBlocksAtDepth(header.ParentHash)
+	equivocated, err := b.verifyBlockEquivocation(header)
+	if err != nil {
+		return fmt.Errorf("could not verify block equivocation for header %s: %w", header.Hash(), err)
+	}
+	if equivocated {
+		return ErrProducerEquivocated
+	}
 
-	for _, currentHash := range hashes {
-		currentHeader, err := b.blockState.GetHeader(currentHash)
-		if err != nil {
-			return fmt.Errorf("failed get header %s", err)
-		}
+	return nil
+}
 
-		currentBlockProducerIndex, err := getAuthorityIndex(currentHeader)
-		if err != nil {
-			return fmt.Errorf("failed to get authority index %s", err)
-		}
+// verifyBlockEquivocation checks if given block's author has occupied corresponding slot more than once.
+// It returns true if block was equivocated.
+func (b *verifier) verifyBlockEquivocation(header *types.Header) (bool, error) {
+	author, err := getAuthorityIndex(header)
+	if err != nil {
+		return false, fmt.Errorf("failed to get authority index for %s: %w", header.Hash(), err)
+	}
 
-		if len(currentHeader.Digest.Types) == 0 {
-			return fmt.Errorf("current header missing digest")
+	currentHash := header.Hash()
+	slot, err := types.GetSlotFromHeader(header)
+	if err != nil {
+		return false, fmt.Errorf("failed to get slot for block %s: %w", currentHash, err)
+	}
+
+	blocksInSlot, err := b.blockState.GetBlocksBySlot(slot)
+	if err != nil {
+		return false, fmt.Errorf("failed to get blocks produced in slot %d: %w", slot, err)
```

Comment:
`slot` should be part of the error context set by `GetBlocksBySlot`, not the caller (to avoid repeating it):

```suggestion
		return false, fmt.Errorf("failed to get blocks produced by slot: %w", err)
```

Reply:
https://github.com/ChainSafe/gossamer/pull/2709#discussion_r1007990573